### ISSUE TITLE
Impute sex chromosomes properly

### ISSUE THREAD

**Issue by rwdavies** [rwdavies/STITCH#46]
per-request

**Comment by GuyReeves** [comment]
is there any advise on imputing sex chromosomes  .  So far I have  only done autosomes, is there are way to do Sex chromosomes.

I believe I have the sex of most or all individuals.

my data is XY Male  XX female Drosophila

**Comment by rwdavies** [comment]
Apologies I never got around to writing this. The simplest thing to do is just run all the samples as if it weren't a sex chromosome. This will be fine for XX samples. For XY samples, imputing the X, it will be slightly less accurate, as all recombinations need to be double recombinations, but it should still work. The right thing to do would be to take sex as an argument and allow the XY samples to have only single recombinations, but alas, that's a future goal

**Comment by tdlong** [comment]
I am interested in if people have looked closely at this.

If I take samples that are a mix of males and females and look at genotype calls for the X chromosome, only for very high INFO_SCORE cutoffs (>0.98) are SNPs generally all 0/0 vs 1/1 for males.  At lower INFO_SCORES, I see lots of 0/1 genotypes in males, but that is impossible for much of the X chromosome. I assume the 0/0 & 1/1 calls are really 0 vs 1, as the X is hemizygous in males.

On the other hand for autosomal SNPs we have good independent evidence STITCH calls are very accurate for intermediate frequency SNPs at much more modest INFO_SCORES (e.g, 0.4).  So I do not think this is just a problem with my data.  If I filter SNPs on HWE scores X-linked SNPs are basically all filtered out (which I think makes sense).  But HWE scores are OK for autosomes. So I do think STITCH is getting mixed up in mixed sex samples for X-linked markers.  I wonder if STITCH is getting mixed up because in males the frequencies of haplotypes are p1, p2, ... pn in males and p1^2, p2^2, ... 2p1p2, .... in females.

But perhaps this is just my system.

Have others just tabulated the genotypes in a good male sample for X-linked markers to test that there are very few 0/1's.

**Comment by rwdavies** [comment]
It's been a while since I've thought of this. There's a mode `diploid-inbred` that should probably work if run on only males? I reckon that's worth a shot? It's designed to run on fully inbred diploid chromosomes, e.g. fully inbred lab mice, and basically assumes both sample chromosomes  are the same, which is effectively what you have with the male hemizygous X. 

Perhaps you could run the males and females separately then? There'd be some batch effects but hopefully it would broadly work?

**Comment by tdlong** [comment]
Awesome.  I was going to try running females and males separately for the X and report back.  But this is an even better idea!

**Comment by tdlong** [comment]
I did the experiment.  Females only, males with the diploid-inbred switch, males only (default model).  I filtered the resulting vcf files for info scores >90% and biallelic SNPs.  In the experiment with male samples only but default model there were still 0/1 calls at a high rate.  So that is not working as these genotypes are not correct.  For the diploid-inbred switch, the number of SNPs getting past the high info filter is greatly reduced compared to females.  But most calls are 0/0 or 1/1 (oddly, if I output dosages, some dosages are closest to 1, albeit at a low rate).

Now I have to see about validation.  This is difficult as these are Peromyscus "mice".  But I should be able to validate a subset.

I wonder if anyone has some human male samples that could be tested for the X-chromosome. As then validation is more straightforward if those males have also been SNP chipped??

**Comment by rwdavies** [comment]
Do you have a similar number of males and females? Remember for the males you'll have effectively half the information (both sequenced reads, and sequenced haplotypes, although you have the benefit that the samples should be pre-phased). 0.90 INFO score is a high bar too in my opinion.

The part where you write `But most calls are 0/0 or 1/1 (oddly, if I output dosages, some dosages are closest to 1, albeit at a low rate).`, are you describing a bug? Aren't dosages always output? Dosages and genotypes (and genotype posteriors) have a defined relationship, this does seem weird. 

1000G will definitely have both sequencing data and arrays if you want to take this all the way

**Comment by tdlong** [comment]
Yes, similar number of males and females.  But not a great number, 277 males.  But coverages are approaching 1X and lots of LD, as this is an "diversity outbred panel" type situation.  Although N is low, I find this to be a sweet spot for STITCH.  I agree 0.9 is a high bar, but I thought I would start with really confident calls, and work backward from there.

"diploid-inbred" outputs some dosages near "1" (I have to go in and look at genotypes for those cases I am assuming the dosage reflect a 0/1 genotype).  Allowing 0/1 in inbred lines is not necessarily a bug.  Inbred lines are not always inbred, this may be intended behavior.

I was hoping someone already has bams for sexed individuals and scripts on a server ... and is curious. STITCH pipelines are already subsetting chromosomes, it is easy to flip the "diploid-inbred", a little more work to subet the bams by sex.  It is easy to  look at the rate of heterozygote calls in males at X-linked SNPs.  (This is why I think it wonderful to have a thread in the git!).  It would be a great experiment if someone has some mouse or rat DO samples in their queue.

**Comment by tdlong** [comment]
Here is a quick parser for a vcf to get the distribution of genotype calls for each individual

```bash
# given the output from STITCH
# a quick way to look number of 00,01,11 calls by individual
vcf="??"
zcat $vcf | \
	bcftools view -m 2 -M 2 -v snps -q 0.05:minor | \
	bcftools query -f '%CHROM@%POS, %REF, %ALT[, %DS]\n’ | \
	./count_columns.awk

#################
# count_columns.awk 
#################

#!/usr/bin/awk -f

BEGIN {FS=", "; OFS="\t"}
{
    for (i=4; i<=NF; i++) {
        rounded = sprintf("%.0f", $i)
        counts[i,rounded]++
    }
}
END {
    for (i=4; i<=NF; i++) {
        col0 = (i SUBSEP 0 in counts) ? counts[i,0] : 0
        col1 = (i SUBSEP 1 in counts) ? counts[i,1] : 0
        col2 = (i SUBSEP 2 in counts) ? counts[i,2] : 0
        print "Column " i-3 ": " col0 " " col1 " " col2
    }
}
```

**Comment by RobertDaviesGenomics** [comment]
That's pretty cool, thanks! Yeah as you say, the tricky thing is the potential for bias with two separate runs of STITCH. You'd want to adjust subsequent runs for sex, and hope that was sufficient

Not sure if I could get you or @Zilong-Li  to take a stab at modifying the underlying code base, to do the proper thing, i.e. input a file with sexes, then use the respective method per-sex, unifying the parameter updates (which I think should work)? 

There's also the potential for the simpler option, of just running one sex, say females, then imputing the males using diploid-inbred using the parameters for the females. I think some code I used to have on a private version of STITCH (which I called STITCH_again) was lost unfortunately, but wouldn't be that hard to get something like that going again

**Comment by rwdavies** [comment]
That's pretty cool, thanks! Yeah as you say, the tricky thing is the potential for bias with two separate runs of STITCH. You'd want to adjust subsequent runs for sex, and hope that was sufficient

Not sure if I could get you or @Zilong-Li to take a stab at modifying the underlying code base, to do the proper thing, i.e. input a file with sexes, then use the respective method per-sex, unifying the parameter updates (which I think should work)?

There's also the potential for the simpler option, of just running one sex, say females, then imputing the males using diploid-inbred using the parameters for the females. I think some code I used to have on a private version of STITCH (which I called STITCH_again) was lost unfortunately, but wouldn't be that hard to get something like that going again

**Comment by Deeeeen** [comment]
Hi @rwdavies, @tdlong reached out to us a while back regarding to this GitHub issue. Per @tdlong's request, I took a look at the chrX genotype content on the results of our recent published paper https://doi.org/10.1093/g3journal/jkae291. 

We imputed chrX for 14,505 HS rats (F: 7,222; M: 7,283) using diploid mode (niterations=2) with a reference panel that was built from the founders of this outbred population. Here, I am showing you 2 figures on chrX GT proportion in females and males, 1) STITCH raw output without any filtering, 2) after filtering out SNPs with INFO score <= 0.9. The annotation in red is the mean of each box.

As you can see, in our experiment, the proportion of 0/1 genotype on chrX for males are as expected very low, although there is a huge tail of outliers. In general, the mean is below 0.4%. I can't say for sure if this decent result benefited from the large sample size or the reference panel, but I hope this information helps with the problem.

Please feel free to let me know what you think.

![Image](https://github.com/user-attachments/assets/8e1c0c44-1bef-43e6-be3d-ae1bf85f7e5f)
![Image](https://github.com/user-attachments/assets/3947902e-3869-4f25-bb2a-001c0a730da0)
